### ISSUE TITLE
Specification of expectations for English itwêwina search (single terms)

### ISSUE THREAD

**Issue by aarppe** [UAlbertaALTLab/cree-intelligent-dictionary#543]
1. When `crkeng.xml` is imported, each token/word in the English translations should be indexed as follows:
a. as-is, without any initial or final punctuation: _Edmonton_ -> _Edmonton_
b. transformed into lower case, without any initial or final punctuation: _Edmonton_ -> _edmonton_
c. transformed into lower case and with token-internal punctuation as well as initial/final punctuation removed: _beaver-berries_ -> _beaverberries_
d. stemmed using whatever algorithm we have available: _beaver-berries_ -> _beaver-berry_ (or _beaver-berri_ or _beaver-berrie_)
e. For compound words, i.e. ones with hyphens, the components should be indexed according to steps `a-e`: _beaver-berries_ -> _beaver + berries_ -> ...

| Input | Processing | Index |
| ---     | ---              | ---     |
| Edmonton | as-is | Edmonton |
| beaver-berries | as-is | beaver-berries |
| Edmonton | lower-case | edmonton |
| beaver-berries | no punctuation | beaverberries |
| beaver-berries | compound components | beaver, berries |
| beaver-berries | stemming | beaver-berri |
| beaverberries | stemming | beaverberri |
| beaver | stemming | beaver |
| berries | stemming | berri |

2. When an English search is initiated, the search string should be processed as follows:
a. as-is: _Calgary_ matches _Calgary_; _beaver-berries_ matches _beaver-berries_
b. transformed into lower case: _Calgary_ -> _calgary_ matches _calgary_ (which co-indexes _Calgary_)
c. transformed into lower case and with token-internal punctuation removed: _beaver-berries_ -> _beaverberries_ matches _beaverberries_ (which co-indexes _beaver-berries_)
d. stemmed: _beaver-berries_ -> _beaver-berri_ matches _beaver-berri_ (which co-indexes _beaber-berries_)
e. prefix search: any of _b_, _be_, _bea_, _beav_, _beave_, and _beaver_ match _beaver-berries_ + _beaverberries_ + _beaver_ (which co-index for _beaver-berries_); _berries_ -> any of _b_, _be_, _ber_, _berr_, and _berri_ match _beaver-berri_ (co-indexing with beaver-berries); BUT I don't know how we would get from _berry_ to _beaver-berries_ without a proper stemmer that is able to stem _berries_ as _berry_ (of course _berr_ will match _berri_).

| Search | Processing | Index | Match |
| ---        | ---              | ---     | ---      |
| Calgary | as-is | Calgary | Calgary |
| beaver-berries | as-is | beaver-berries | beaver-berries |
| Calgary | lower-case | calgary | Calgary |
| beaver-berries | strip punctuation | beaverberries | beaver-berries |
| beaver-berries | stemming | beaver-berri | beaver-berries |
| b, be, bea, beav, beave, beaver | prefix search | beaver | beaver-berries |
| b, be, ber, berr, berri, berrie | prefix search | berries | beaver-berries |

**[edit -- added this bridging case]**
As a bridge case combining this issue with #589, searching with the word pair _working together_, the following is what would be expected.

a. as-is: _working together_ matches _working_ and _together_.
d. stemmed : _working together_ is stemmed as _work_ and _together_, and should match the intersection of matches with these stems searched separately.
e. prefix search: the last space-separated word might allow for prefix search (perhaps after the first 2-3 letters).

| Search | Processing | Index | Match |
| ---        | ---              | ---     | ---      |
| working together | stemming | work + together | work AND together |
| w, wo, wor, work | prefix search | work | working |
| worki, workin | prefix search | working | working |
| work t, work to, work tog, work toge, work toget, work togeth, work togethe | prefix search | work + together | work AND together |

**Comment by aarppe** [comment]
@Madoshakalaka The above behavior is what I'd expect to happen when crkeng.xml is imported and its English translations are used to index the Cree dictionary entries, as well as how the English search component will process the search string.

**Comment by aarppe** [comment]
Here's some GAWK code that would seem to implement 1. Indexing above. The stemmer is the crappiest ever, but perhaps good enough for our purposes.

```
echo "-beaver-berries'" |
gawk '{ sub("^[[:punct:]]*","",$0); sub("[[:punct:]]*$","",$0); nw=split($0,w,"[- ]"); print $0; for(i=1; i<=nw; i++) print w[i]; }' |
gawk '{ print $0; if(tolower($0)!=$0) print tolower($0); }' |
gawk '{ stem=$0; sub("(((e)?s)|(ed)|(n)|(ing)|(er)|(est))$","",stem); print $0; if(stem!=$0) print stem; }'

beaver-berries
beaver-berri
beaver
beav
berries
berri
```

**Comment by aarppe** [comment]
Addressed via #784.
